# Actividad evaluada Semana 2 — Cuaderno con funciones para procesamiento funcional de datos

**Bootcamp:** AI Machine Learning Trainee 1 (AIMLF1) · **KODIGO**
**Módulo 1:** Fundamentos de Python para Análisis de Datos

---

## 👤 Datos del estudiante (obligatorio — edita esta celda)

| Campo | Respuesta |
|---|---|
| **Nombre completo** | Alejandro José Orellana Reyes |
| **Fecha de entrega** | 22/08/2026 |

---

## 🧩 Problemática

Una empresa de análisis de datos recibe información de múltiples fuentes en formato de listas. Sin embargo, estos datos contienen inconsistencias: **valores nulos, duplicados, tipos de datos incorrectos y formatos heterogéneos**. Tu misión es **automatizar el proceso de limpieza y transformación** para garantizar la calidad de los análisis, aplicando estructuras de control, funciones parametrizadas, expresiones lambda y manejo de excepciones.

## 📋 Instrucciones generales

1. Trabaja **sobre esta plantilla**: completa las celdas marcadas con `# 💻 TU CÓDIGO AQUÍ` y las celdas Markdown marcadas con ✍️. No borres los enunciados ni cambies el orden de las secciones.
2. **Nombra tus funciones como se sugiere en cada enunciado** (facilita la revisión); puedes agregar funciones extra si lo deseas.
3. **Requisitos técnicos obligatorios** (cuenta para la nota):
   - Mínimo **8 funciones parametrizadas** en total.
   - **Todas** las funciones con **docstring** (resumen + `Args:` + `Returns:`; agrega `Raises:` donde aplique).
   - **`try`-`except` obligatorio** en las funciones de validación (Sección 2).
   - **Al menos 2 funciones** que usen expresiones **lambda**.
   - Código **comentado** explicando la lógica y **celdas Markdown** explicando cada sección.
4. En cada sección **muestra el antes y el después** de aplicar tus funciones (imprime muestras) y **demuestra el manejo de errores** con al menos un dato inválido.
5. Antes de entregar: **Kernel → Restart & Run All**. El cuaderno debe correr completo, de arriba abajo, **sin errores** (los errores deben quedar *capturados*, nunca sueltos).
6. Sube a Moodle el archivo con el nombre: **`Apellido_Nombre_Procesamiento_Datos.ipynb`**.

**Puntaje total: 10.0 puntos** — la rúbrica detallada está al final del cuaderno.

---

## 📦 Paso 0 · Carga de los datos (no modificar)

La celda siguiente descarga el dataset oficial de la actividad ([Kodigo-academic/basic_sample_data](https://github.com/Kodigo-academic/basic_sample_data)) e importa el diccionario `datos_empresa` con las 10 listas crudas. **Ejecútala tal cual.**

In [5]:
# ⚙️ CELDA DADA — ejecútala sin modificar
import urllib.request

url = "https://raw.githubusercontent.com/Kodigo-academic/basic_sample_data/main/main.py"
urllib.request.urlretrieve(url, "main.py")

from main import datos_empresa

print("Conjuntos disponibles:", list(datos_empresa.keys()))

✅ Datos de prueba cargados exitosamente
Total de conjuntos de datos: 10
Conjuntos disponibles: ['nombres', 'edades', 'correos', 'salarios', 'productos', 'cantidades', 'fechas', 'codigos_postales', 'calificaciones', 'departamentos']


---
Implementa **cuatro funciones** (bucles + condicionales; aquí no se usan excepciones):

1. `eliminar_nulos_y_vacios(lista)` — retorna una lista **nueva** sin `None` y sin cadenas vacías (incluyendo las que quedan vacías tras quitar espacios, como `"  "`).
2. `quitar_espacios(lista)` — retorna la lista con `.strip()` aplicado a los elementos que sean strings (los no-string pasan intactos).
3. `normalizar_texto(lista, formato="titulo")` — retorna la lista con los strings normalizados; el parámetro `formato` acepta `"titulo"`, `"minusculas"` o `"mayusculas"`.
4. `eliminar_duplicados(lista)` — retorna la lista sin duplicados, **manteniendo el primer valor encontrado** y el orden.

Luego **pruébalas** sobre `datos_empresa["nombres"]` y `datos_empresa["productos"]` mostrando el antes y el después (cantidad de elementos y una muestra).

Esta sección es para crear una lista con los valores limpios, eliminando la "basura" que sería toda aquella información que dificulta o incluso podría romper el programa por contener espacios en blanco, saltos de línea, formateo de mayusculas minusculas.

In [6]:


def eliminar_nulos_y_vacios(lista):
    resultado = []
    for elemento in lista:
        # Saltar None directamente
        if elemento is None:
            continue
        # Si es string, verificar que no quede vacío tras strip
        if isinstance(elemento, str) and elemento.strip() == "":
            continue
        resultado.append(elemento)
    return resultado


def quitar_espacios(lista):
    resultado = []
    for elemento in lista:
        if isinstance(elemento, str):
            resultado.append(elemento.strip())
        else:
            resultado.append(elemento)
    return resultado


def normalizar_texto(lista, formato="titulo"):
    resultado = []
    for elemento in lista:
        if isinstance(elemento, str):
            if formato == "titulo":
                resultado.append(elemento.title())
            elif formato == "minusculas":
                resultado.append(elemento.lower())
            elif formato == "mayusculas":
                resultado.append(elemento.upper())
            else:
                resultado.append(elemento)  # formato desconocido, dejar tal cual
        else:
            resultado.append(elemento)
    return resultado


def eliminar_duplicados(lista):
    vistos = set()
    resultado = []
    for elemento in lista:
        # Usar una tupla para hashear listas anidadas si las hay, pero aquí son strings/None
        if elemento not in vistos:
            vistos.add(elemento)
            resultado.append(elemento)
    return resultado

In [7]:

# Prueba con nombres
nombres_crudo = datos_empresa["nombres"]
print("PRUEBA nombres:", len(nombres_crudo), "->", nombres_crudo[:8])

# Encadenar funciones: quitar espacios, normalizar, eliminar nulos/vacíos, duplicados
nombres_limpios = eliminar_duplicados(
    eliminar_nulos_y_vacios(normalizar_texto(quitar_espacios(nombres_crudo), "titulo"))
)
print("DESPUÉS:", len(nombres_limpios), "->", nombres_limpios[:8])

# Prueba con productos
productos_crudo = datos_empresa["productos"]
print("PRUEBA productos:", len(productos_crudo), "->", productos_crudo[:8])

productos_limpios = eliminar_duplicados(
    eliminar_nulos_y_vacios(normalizar_texto(quitar_espacios(productos_crudo), "titulo"))
)
print("DESPUÉS:", len(productos_limpios), "->", productos_limpios[:8])


PRUEBA nombres: 27 -> ['  Juan Pérez  ', 'maría garcía', 'CARLOS LÓPEZ', None, 'Ana Martínez', 'juan pérez', '', '  Pedro Sánchez']
DESPUÉS: 19 -> ['Juan Pérez', 'María García', 'Carlos López', 'Ana Martínez', 'Pedro Sánchez', 'Luis Rodríguez', 'Carmen Flores', 'Laura Ruiz']
PRUEBA productos: 27 -> ['laptop', 'mouse', 'TECLADO', 'laptop', None, '  monitor  ', 'Mouse', 'audifonos']
DESPUÉS: 13 -> ['Laptop', 'Mouse', 'Teclado', 'Monitor', 'Audifonos', 'Webcam', 'Tablet', 'Impresora']


---
## Sección 2 · Funciones de validación con manejo de excepciones — *2.5 pt*

La estrategia EAFP (*Easier to Ask for Forgiveness than Permission*) consiste en intentar realizar una operación directamente y capturar las excepciones si llega a fallar, en lugar de verificar previamente cada condición. Permite manejar errores de conversión y formato, evitando que el programa colapse.

Implementa **cinco funciones validadoras**. Todas deben usar **`try`-`except`** (nunca dejar escapar un error), retornar el **valor convertido** cuando el dato es recuperable y **`None`** cuando no lo es:

1. `validar_edad(dato)` — convierte a entero y acepta solo el rango **0–120**.
2. `validar_correo(dato)` — limpia espacios, pasa a minúsculas y acepta solo si contiene **una** `@` con texto antes, y un dominio con al menos un punto y texto a ambos lados (rechaza `"usuario@"`, `"luis@email"`, `"isabel@.com"`).
3. `validar_salario(dato)` — convierte a `float` y acepta solo **valores positivos**.
4. `validar_calificacion(dato)` — convierte a entero y acepta solo el rango **0–100**.
5. `validar_codigo_postal(dato)` — limpia espacios y acepta solo **exactamente 4 dígitos numéricos** (retórnalo como string, p. ej. `"1101"`).

Luego **pruébalas** sobre sus listas correspondientes de `datos_empresa`, y **demuestra el manejo de errores**: muestra al menos 3 datos inválidos de cada lista y cómo tu función los neutraliza retornando `None` en lugar de colapsar.


In [8]:

import re

def validar_edad(dato):
    #Valida y convierte un dato a entero en el rango 0–120.
    try:
        edad = int(dato)
        if 0 <= edad <= 120:
            return edad
        return None
    except (ValueError, TypeError):
        return None


def validar_correo(dato):
    #Valida un correo electrónico limpiando espacios y pasando a minúsculas.
    try:
        correo = str(dato).strip().lower()
        # Debe contener exactamente una @
        if correo.count("@") != 1:
            return None
        parte_local, dominio = correo.split("@")
        # Texto antes de la @
        if not parte_local:
            return None
        # Dominio debe tener al menos un punto con texto a ambos lados
        if "." not in dominio:
            return None
        # Verificar que no empiece ni termine con punto y haya texto entre puntos
        partes_dominio = dominio.split(".")
        if any(not parte for parte in partes_dominio):
            return None
        return correo
    except (ValueError, TypeError, AttributeError):
        return None


def validar_salario(dato):
    #Valida y convierte un dato a float positivo.
    try:
        salario = float(dato)
        if salario > 0:
            return salario
        return None
    except (ValueError, TypeError):
        return None


def validar_calificacion(dato):
    #Valida y convierte un dato a entero en el rango 0–100.
    try:
        calif = int(dato)
        if 0 <= calif <= 100:
            return calif
        return None
    except (ValueError, TypeError):
        return None


def validar_codigo_postal(dato):
    #Valida que el dato sea exactamente 4 dígitos numéricos.
    try:
        cp = str(dato).strip()
        if re.fullmatch(r"\d{4}", cp):
            return cp
        return None
    except (ValueError, TypeError, AttributeError):
        return None


In [9]:

def probar_validador(lista, validador, nombre):
    #Aplica un validador a una lista y muestra estadísticas + ejemplos de errores
    validados = [validador(x) for x in lista]
    validos = [v for v in validados if v is not None]
    rechazados = [lista[i] for i in range(len(lista)) if validados[i] is None]
    print(f"VALIDADOR: {nombre} | válidos: {len(validos)} | rechazados: {len(rechazados)}")
    print("  3 inválidos neutralizados:", rechazados[:3], "-> None")
    return validados


# Aplicar a cada lista
edades_validadas = probar_validador(datos_empresa["edades"], validar_edad, "validar_edad")
correos_validados = probar_validador(datos_empresa["correos"], validar_correo, "validar_correo")
salarios_validados = probar_validador(datos_empresa["salarios"], validar_salario, "validar_salario")
calificaciones_validadas = probar_validador(datos_empresa["calificaciones"], validar_calificacion, "validar_calificacion")
codigos_validados = probar_validador(datos_empresa["codigos_postales"], validar_codigo_postal, "validar_codigo_postal")


VALIDADOR: validar_edad | válidos: 18 | rechazados: 9
  3 inválidos neutralizados: [None, -5, 150] -> None
VALIDADOR: validar_correo | válidos: 13 | rechazados: 14
  3 inválidos neutralizados: ['correo_invalido', None, ''] -> None
VALIDADOR: validar_salario | válidos: 18 | rechazados: 9
  3 inválidos neutralizados: [None, 'abc', -500] -> None
VALIDADOR: validar_calificacion | válidos: 18 | rechazados: 9
  3 inválidos neutralizados: [None, 105, -10] -> None
VALIDADOR: validar_codigo_postal | válidos: 14 | rechazados: 13
  3 inválidos neutralizados: [None, 'ABCD', ''] -> None


---
## Sección 3 · Funciones de transformación con lambda — *1.5 pt*

Una función lambda es una función de una sola expresión que se define en línea con lambda argumentos: expresión, sin necesidad de def ni return explícito. Se usa cuando la lógica es breve y se necesita pasar una función como argumento

Implementa **al menos tres transformaciones**; **mínimo dos deben usar `lambda`** (con `map()`, `filter()` o `sorted()`):

1. `convertir_a_tipo(lista, tipo)` — recibe una lista de valores mixtos y un tipo destino (`int` o `float`); usa `map()` con una lambda que convierta lo convertible y retorna solo los valores convertidos.
2. `ajustar_salarios(lista_salarios, porcentaje)` — aplica un incremento (o descuento si es negativo) porcentual a una lista de salarios **ya limpios**, usando `map()` + `lambda`. Redondea a 2 decimales.
3. `categorizar_edades(lista_edades)` — usa `map()` + `lambda` + operador ternario para etiquetar edades **ya limpias**: `"joven"` (< 30), `"adulto"` (30–59), `"adulto mayor"` (60+).
4. `filtrar_por_criterio(lista, criterio)` — recibe una función/lambda como parámetro `criterio` y usa `filter()` para retornar los elementos que lo cumplen (pruébala filtrando productos que empiecen con una letra, o departamentos iguales a `"Ventas"`).

**Prueba** cada transformación mostrando entrada y salida.


In [10]:

from functools import reduce


def convertir_a_tipo(lista, tipo):
    #Convierte a un tipo destino los valores convertibles de una lista.

    def _convertir(x):
        try:
            if tipo == "int":
                v = int(float(x)) if isinstance(x, str) else int(x)
                return v if v >= 0 else None
            elif tipo == "float":
                v = float(x)
                return v if v >= 0 else None
            return None
        except (ValueError, TypeError):
            return None

    # map() con lambda que aplica la conversión segura
    resultado = list(filter(lambda v: v is not None, map(lambda x: _convertir(x), lista)))
    return resultado


def ajustar_salarios(lista_salarios, porcentaje):
    # Aplica un cambio porcentual a salarios ya limpios.

    # map() + lambda: nuevo_salario = salario * (1 + porcentaje/100)
    return list(map(lambda s: round(s * (1 + porcentaje / 100), 2), lista_salarios))


def categorizar_edades(lista_edades):
    # Etiqueta edades ya limpias según su categoría de edad.
    # map() + lambda + operador ternario
    return list(map(
        lambda e: "joven" if e < 30 else ("adulto" if e <= 59 else "adulto mayor"),
        lista_edades
    ))


def filtrar_por_criterio(lista, criterio):
    #Retorna los elementos de la lista que cumplen el criterio dado.
    return list(filter(criterio, lista))


In [11]:

mezcla = ["10", 3.7, "abc", None, " -2 ", 8, "15.9"]
print("1) convertir_a_tipo -> int:", convertir_a_tipo(mezcla, "int"))
print("1) convertir_a_tipo -> float:", convertir_a_tipo(mezcla, "float"))

# Ajustar_salarios: incremento del 10% (usa salarios válidos)
salarios_limpios = [s for s in salarios_validados if s is not None]
print("2) ajustar_salarios (+10%):", ajustar_salarios(salarios_limpios, 10)[:8])

# Categorizar_edades (usa edades válidas)
edades_limpias = [e for e in edades_validadas if e is not None]
print("3) categorizar_edades:", categorizar_edades(edades_limpias)[:8])

# Filtrar_por_criterio: productos que empiecen con 'L'
productos_txt = [p for p in productos_limpios if isinstance(p, str)]
print("4) filtrar_por_criterio:", filtrar_por_criterio(productos_txt, lambda x: x.lower().startswith("l")))


1) convertir_a_tipo -> int: [10, 3, 8, 15]
1) convertir_a_tipo -> float: [10.0, 3.7, 8.0, 15.9]
2) ajustar_salarios (+10%): [1650.55, 2200.0, 2750.83, 1980.0, 3300.0, 4950.28, 3520.55, 3080.0]
3) categorizar_edades: ['joven', 'adulto', 'adulto', 'joven', 'adulto', 'joven', 'adulto', 'adulto']
4) filtrar_por_criterio: ['Laptop']


---
## Sección 4 · Procesamiento con estructuras de control — *2.0 pt*

En esta sección se convina todo lo anterior. Primero se genera un reporte de calidad que recorre cada conjunto de datos_empresa y cuenta cuántos registros son válidos e inválidos aplicando la función de limpieza correspondiente. Construimos datos_limpios, un nuevo diccionario con la versión limpia de cada lista. Y finalmente calculo **estadísticas básicas** (mínimo, máximo y promedio) sobre edades, salarios y calificaciones ya limpias.

Aquí se orquesta todo lo anterior. Implementa:

1. `generar_reporte_calidad(datos)` — itera sobre el diccionario `datos_empresa` con un bucle y retorna/imprime, **por cada conjunto**: total de registros, cuántos son válidos y cuántos inválidos (usa tus funciones de las secciones 1 y 2 según el conjunto).
2. `limpiar_todo(datos)` — retorna un **nuevo diccionario** `datos_limpios` con la versión limpia de todas las listas (aplica a cada conjunto las funciones que le correspondan: texto → limpieza básica; numéricos y correos → validadores).
3. `estadisticas_basicas(lista_numerica)` — retorna (mínimo, máximo, promedio) de una lista numérica ya limpia. Aplícala al menos a **edades**, **salarios** y **calificaciones** limpias, presentando los resultados con f-strings formateadas.


In [12]:

LIMPIADORES = {
    "nombres":          lambda lista: eliminar_duplicados(eliminar_nulos_y_vacios(normalizar_texto(quitar_espacios(lista), "titulo"))),
    "productos":        lambda lista: eliminar_duplicados(eliminar_nulos_y_vacios(normalizar_texto(quitar_espacios(lista), "titulo"))),
    "departamentos":    lambda lista: eliminar_duplicados(eliminar_nulos_y_vacios(normalizar_texto(quitar_espacios(lista), "titulo"))),
    "edades":           lambda lista: [validar_edad(x) for x in lista],
    "salarios":         lambda lista: [validar_salario(x) for x in lista],
    "calificaciones":   lambda lista: [validar_calificacion(x) for x in lista],
    "codigos_postales": lambda lista: [validar_codigo_postal(x) for x in lista],
    "correos":          lambda lista: [validar_correo(x) for x in lista],
    "cantidades":       lambda lista: convertir_a_tipo(lista, "int"),
    "fechas":           lambda lista: eliminar_nulos_y_vacios(quitar_espacios(lista)),
}


def generar_reporte_calidad(datos):
    # Genera e imprime un reporte de calidad por cada conjunto de datos.
    reporte = {}
    print("=" * 60)
    print("REPORTE DE CALIDAD DE DATOS")
    print("=" * 60)
    print(f"{'Conjunto':<16} {'Total':>6} {'Válidos':>8} {'Inválidos':>9} {'% Rechazo':>10}")
    print("-" * 60)

    for clave, lista in datos.items():
        limpiador = LIMPIADORES[clave]
        # Limpiar y contar cuántos valores quedaron válidos (no None / no vacíos)
        limpios = limpiador(lista)
        validos = len(limpios)
        total = len(lista)
        invalidos = total - validos
        pct = (invalidos / total * 100) if total else 0
        reporte[clave] = (total, validos, invalidos)
        print(f" {clave:<15} {total:>6} {validos:>8} {invalidos:>9} {pct:>9.1f}%")

    print("-" * 60)
    return reporte


def limpiar_todo(datos):
    # Construye un nuevo diccionario con la versión limpia de cada lista.

    datos_limpios = {}
    for clave, lista in datos.items():
        limpiador = LIMPIADORES[clave]
        # Filtrar valores inválidos (None) para dejar solo datos utilizables
        limpios = limpiador(lista)
        datos_limpios[clave] = [v for v in limpios if v is not None]
    return datos_limpios


def estadisticas_basicas(lista_numerica):
    # Calcula mínimo, máximo y promedio de una lista numérica ya limpia.
    if not lista_numerica:
        return (0, 0, 0.0)
    minimo = min(lista_numerica)
    maximo = max(lista_numerica)
    promedio = sum(lista_numerica) / len(lista_numerica)
    return (minimo, maximo, promedio)


In [13]:

# Reporte de calidad: válidos vs inválidos por conjunto
reporte = generar_reporte_calidad(datos_empresa)

# Construir el diccionario de datos limpios
datos_limpios = limpiar_todo(datos_empresa)
print("DATOS LIMPIOS:", {k: len(v) for k, v in datos_limpios.items()})

# Estadísticas básicas: edades, salarios y calificaciones limpias
for nombre, lista in [
    ("Edades", datos_limpios["edades"]),
    ("Salarios", datos_limpios["salarios"]),
    ("Calificaciones", datos_limpios["calificaciones"]),
]:
    minimo, maximo, promedio = estadisticas_basicas(lista)
    print(f"{nombre}: min={minimo}, max={maximo}, promedio={promedio:.2f}")


REPORTE DE CALIDAD DE DATOS
Conjunto          Total  Válidos Inválidos  % Rechazo
------------------------------------------------------------
 nombres             27       19         8      29.6%
 edades              27       27         0       0.0%
 correos             27       27         0       0.0%
 salarios            27       27         0       0.0%
 productos           27       13        14      51.9%
 cantidades          27       20         7      25.9%
 fechas              27       21         6      22.2%
 codigos_postales     27       27         0       0.0%
 calificaciones      27       27         0       0.0%
 departamentos       27       12        15      55.6%
------------------------------------------------------------
DATOS LIMPIOS: {'nombres': 19, 'edades': 18, 'correos': 13, 'salarios': 18, 'productos': 13, 'cantidades': 20, 'fechas': 21, 'codigos_postales': 14, 'calificaciones': 18, 'departamentos': 12}
Edades: min=19, max=60, promedio=34.78
Salarios: min=1200.0, ma

---
## Sección 5 · Conclusiones — *dentro del puntaje de documentación*

Al revisar la calidad de los datos, encontré que muchos registros llegaron incompletos o con errores. El conjunto más sucio fue **departamentos**, con un 55.6% de registros rechazados, seguido de cerca por **productos** con 51.9%; en ambos casos los problemas principales fueron valores repetidos, espacios de más y celdas vacías. **nombres** tiene (29.6%), **cantidades** (25.9%) y **fechas** (22.2%) presentaron errores de formato y valores nulos. Sin embargo, listas como edades, salarios, calificaciones y códigos postales estuvieron mucho más limpias al recibirlas.

Estos números me indican que las principales fuentes del problema son la captura manual de datos y la falta de reglas claras al ingresarlos. Por eso le recomendaría aplicar validaciones desde el momento en que se registra la información ayudaría a reducir los rechazos y a tener análisis más confiables.


---
## 📊 Rúbrica de evaluación (total: 10.0 puntos)

| # | Criterio | Puntos | Se evalúa |
|---|---|---|---|
| 1 | Limpieza básica | 2.0 | Las 4 funciones correctas (0.4 c/u = 1.6); pruebas antes/después sobre nombres y productos (0.4) |
| 2 | Validación con excepciones | 2.5 | Las 5 validadoras correctas con try-except y retorno de `None` (0.4 c/u = 2.0); demostración de manejo de errores con datos inválidos (0.5) |
| 3 | Transformación con lambda | 1.5 | Al menos 3 transformaciones correctas (1.0); uso genuino de lambda en ≥ 2 de ellas con map/filter/sorted (0.5) |
| 4 | Procesamiento y reportes | 2.0 | Iteración sobre `datos_empresa` con reporte válidos vs inválidos (0.75); `datos_limpios` completo (0.75); estadísticas básicas formateadas (0.5) |
| 5 | Documentación | 1.5 | Docstrings completos en TODAS las funciones (0.75); celdas Markdown explicativas por sección + conclusiones con números (0.75) |
| 6 | Calidad global | 0.5 | Mínimo 8 funciones parametrizadas, nombres descriptivos en snake_case, código comentado, f-strings |

**Penalizaciones:**

- El cuaderno **no ejecuta completo** con *Restart & Run All*: **−1.0 punto**.
- Sin nombre del estudiante en el encabezado: **−0.25 puntos**.
- Celdas de enunciado borradas o desordenadas: **−0.25 puntos**.
- Menos de 8 funciones parametrizadas o funciones sin docstring: se descuenta dentro de los criterios 5 y 6.

---

*Declaro que esta entrega es trabajo propio, elaborado individualmente para el bootcamp AIMLF1 de KODIGO.*